tests/msc2716_test.go

```diff
@@ -0,0 +1,643 @@
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+
+			backfillBatchHistoricalMessages(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				"$some-non-existant-event-id",
+				time.Now(),
+				"",
+				1,
+				// Status
+				// TODO: Seems like this makes more sense as a 404
+				// But the current Synapse code around unknown prev events will throw ->
+				// `403: No create event in auth events`
+				403,
+			)
+		})
+
+		t.Run("Normal users aren't allowed to backfill messages", func(t *testing.T) {
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+			alice.JoinRoom(t, roomID, nil)
+
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 1)
+			eventBefore := eventsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			backfillBatchHistoricalMessages(
+				t,
+				alice,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore,
+				"",
+				1,
+				// Status
+				// Normal user alice should not be able to backfill messages
+				403,
+			)
+		})
+
+		t.Run("TODO: Test if historical avatar/display name set back in time are picked up on historical messages", func(t *testing.T) {
+			t.Skip("Skipping until implemented")
+			// TODO: Try adding avatar and displayName and see if historical messages get this info
+		})
+
+		t.Run("Historical messages are visible when joining on federated server", func(t *testing.T) {
+			t.Skip("Skipping until federation is implemented")
+			t.Parallel()
+
+			roomID := as.CreateRoom(t, struct{}{})
+			alice.JoinRoom(t, roomID, nil)
+
+			eventsBefore := createMessagesInRoom(t, alice, roomID, 1)
+			eventBefore := eventsBefore[0]
+			timeAfterEventBefore := time.Now()
+
+			// eventsAfter
+			createMessagesInRoom(t, alice, roomID, 3)
+
+			// Register and join the virtual user
+			ensureVirtualUserRegistered(t, as, virtualUserLocalpart)
+
+			backfillRes := backfillBatchHistoricalMessages(
+				t,
+				as,
+				virtualUserID,
+				roomID,
+				eventBefore,
+				timeAfterEventBefore,
+				"",
+				2,
+				// Status
+				200,
+			)
+			_, historicalEvents := getEventsFromBatchSendResponse(t, backfillRes)
+
+			// Join the room from a remote homeserver after the backfilled messages were sent
+			remoteCharlie.JoinRoom(t, roomID, []string{"hs1"})
+
+			messagesRes := remoteCharlie.MustDoFunc(t, "GET", []string{"_matrix", "client", "r0", "rooms", roomID, "messages"}, client.WithContentType("application/json"), client.WithQueries(url.Values{
+				"dir":   []string{"b"},
+				"limit": []string{"100"},
+			}))
+
+			must.MatchResponse(t, messagesRes, match.HTTPResponse{
+				JSON: []match.JSON{
+					match.JSONCheckOffAllowUnwanted("chunk", []interface{}{historicalEvents[0], historicalEvents[1]}, func(r gjson.Result) interface{} {
+						return r.Get("event_id").Str
+					}, nil),
+				},
+			})
+		})
+
+		t.Run("Historical messages are visible when already joined on federated server", func(t *testing.T) {
```

Comment:
A lot of these tests have a ton of boilerplate to:
 - Make a few events
 - Insert backfilled events
 - Do the interesting thing

It would be far better to have a single test do:
  - Make a few events
 - Insert backfilled events

Then have sub-tests which execute afterwards which do things like this. It would probably cut >100 lines of code..

Reply:
I've tried refactoring the tests into a single setup with as many of the other tests re-using it. It does require having some of the extra users scattered in the setup like `remoteCharlieAlreadyJoined` and `remoteCharlieAlreadyJoined` instead of in their particular test but that seems fine. You can see how it would look like in https://github.com/matrix-org/complement/blob/dcd84ad906370762bde6444a95d79f8e0ef04772/tests/msc2716_test.go

I think we would also have to remove the `t.Parallel()` stuff though and maintain a certain test order since we rely on certain users not having any scrollback in the room and we would run into some race conditions with whichever test decides to finish first.

I'd prefer to keep the tests independent with clean state for each one.

Reply:
Fair enough